## Introduction

Reliable analysis depends on having clean, consistent, and well-structured data. This notebook focuses on preparing the **Superstore dataset** for exploratory and business analysis.

The primary objective is to **inspect, clean, validate, and standardize the dataset** while preserving the underlying meaning of the data. The cleaning process includes identifying missing values, investigating whether missingness is random or structural, standardizing formats and categorical values, and validating data consistency.

Rather than automatically modifying every irregularity, each data-quality issue is investigated and a decision is made based on the evidence available in the dataset.

The cleaned and validated dataset produced from this notebook will serve as the foundation for the subsequent **Exploratory Data Analysis (EDA) and profitability analysis**.


In [1]:
#Import the necessary libraries
import pandas as pd

In [2]:
# Load source dataset
df = pd.read_csv('ecommerce_sales_customer_analytics_150k.csv')

Data audit and cleaning

In [3]:
# 1. Show all columns instead of truncating with...
pd.set_option('display.max_columns', None)

# 2. Get dataset size
print("Number of rows:", df.shape[0])
print("Number of columns:", df.shape[1])
print("-" * 50) # just a separator line

# 3. Preview the first 5 rows with all columns visible
df.head()

Number of rows: 138116
Number of columns: 46
--------------------------------------------------


,order_id,order_date,order_time,order_status,sales_channel,customer_id,customer_name,customer_age,gender,customer_segment,customer_type,customer_city,customer_state,customer_country,region,customer_postal_code,payment_method,payment_status,currency,shipping_method,warehouse,delivery_days,estimated_delivery_days,delivery_status,return_status,return_reason,customer_rating,review_sentiment,customer_review,marketing_channel,campaign_name,coupon_code,loyalty_points_earned,loyalty_points_redeemed,quantity,gross_sales,discount_amount,tax_amount,shipping_cost,net_sales,product_cost,profit,profit_margin_percentage,customer_lifetime_value,is_repeat_customer,customer_order_count
0,ORD-301242,2023-11-06,16:37:47,Completed,Mobile App,CUST-003102,Jasmine Ryan,55,Male,Consumer,Loyal,Lake Williamberg,Texas,USA,South,86040,Digital Wallet,Paid,USD,Standard,WH-003,3.0,3.0,On Time,NaN,NaN,3.5,Positive,Satisfied with the purchase.,Direct,Default_Campaign,NaN,94,44,5,1350.19,477.89,61.07,12.03,945.40,588.33,345.04,36.50,12459.68,True,11
1,ORD-773460,2025-12-24,01:22:36,Completed,Website,CUST-003124,Scott Chase,26,Male,Premium,Loyal,Kristyport,Baden-Württemberg,Germany,South,62538,Debit Card,Pending,EUR,Economy,WH-005,10.0,10.0,On Time,NaN,NaN,3.6,Positive,Good value for money.,Direct,Default_Campaign,NaN,201,171,8,3144.64,1456.06,320.83,9.00,2018.41,2031.88,-22.47,-1.11,13032.48,True,11
2,ORD-449374,2021-07-05,14:24:21,Completed,Mobile App,CUST-012496,Marc Wheeler,69,Female,Premium,Loyal,Singletonhaven,New York,USA,East,19993,Cash on Delivery,Paid,USD,Express,WH-015,3.0,3.0,On Time,NaN,NaN,4.3,Positive,Good product. Works as expected.,YouTube,NaN,NaN,46,16,5,522.81,97.30,29.79,13.05,468.35,257.73,197.57,42.18,6159.44,True,6
3,ORD-567636,2023-01-21,07:20:26,Completed,Social Media,CUST-023928,Jennifer Smith,65,Male,Consumer,Loyal,Wigginsstad,North Carolina,USA,South,66329,Debit Card,Paid,USD,Express,WH-010,2.0,2.0,On Time,NaN,NaN,3.5,Positive,Satisfied with the purchase.,Email Marketing,NaN,NaN,54,10,2,544.62,53.34,34.39,20.85,546.52,277.84,247.83,45.35,5638.30,True,7
4,ORD-820028,2022-05-13,09:46:21,Completed,Social Media,CUST-012730,Jessica Wang,34,Female,Consumer,Loyal,New Michaelton,Gujarat,India,West,77306,Digital Wallet,Paid,INR,Standard,WH-013,6.0,5.0,Delayed,NaN,NaN,3.0,Neutral,Mixed feelings about this purchase.,Direct,NaN,NaN,278,147,6,2474.52,133.63,421.35,23.03,2785.27,1231.88,1530.36,54.94,13240.10,True,8


In [4]:
df.describe() #Describing the data 

,customer_age,customer_postal_code,delivery_days,estimated_delivery_days,customer_rating,loyalty_points_earned,loyalty_points_redeemed,quantity,gross_sales,discount_amount,tax_amount,shipping_cost,net_sales,product_cost,profit,profit_margin_percentage,customer_lifetime_value,customer_order_count
count,138116.000000,138116.000000,113559.000000,113559.000000,113559.000000,138116.000000,138116.000000,138116.000000,138116.000000,138116.000000,138116.000000,138116.000000,138116.000000,138116.000000,138116.000000,138116.000000,138116.000000,138116.000000
mean,45.862681,50210.889310,4.557772,4.298875,3.677015,103.991051,51.934128,5.615244,1375.384176,237.621614,122.462933,22.278080,1282.503575,708.903449,551.322046,45.065997,8971.639047,6.990559
std,16.467676,28569.767779,2.810012,2.594080,0.483672,115.175640,73.217692,3.371289,1291.009554,378.779459,140.939368,16.022739,1169.274669,730.044970,516.300936,13.120357,4430.506588,2.444166
min,18.000000,516.000000,0.000000,0.000000,1.600000,0.000000,0.000000,1.000000,6.310000,0.000000,0.340000,0.000000,5.800000,1.970000,-1281.590000,-59.550000,22.550000,1.000000
25%,32.000000,25509.000000,2.000000,2.000000,3.400000,17.000000,3.000000,3.000000,458.920000,26.327500,34.650000,9.820000,448.200000,207.667500,190.757500,38.880000,5713.570000,5.000000
50%,46.000000,50183.000000,4.000000,4.000000,3.700000,70.000000,25.000000,5.000000,995.730000,103.285000,77.010000,19.190000,950.760000,478.955000,408.395000,46.990000,8392.525000,7.000000
75%,60.000000,74898.000000,7.000000,6.000000,4.000000,151.000000,71.000000,8.000000,1871.200000,277.125000,156.200000,30.750000,1744.582500,953.762500,752.845000,53.770000,11651.030000,9.000000
max,74.000000,99950.000000,14.000000,10.000000,5.000000,1621.000000,1063.000000,24.000000,16051.190000,6759.020000,2091.520000,149.520000,16218.000000,9963.410000,7083.680000,77.590000,30268.440000,20.000000


In [5]:
# Standardize order date
df['order_date'] = pd.to_datetime(
    df['order_date'],
    format='%Y-%m-%d'
)

# Standardize order time
df['order_time'] = pd.to_datetime(
    df['order_time'],
    format='%H:%M:%S'
).dt.time

In [6]:
# Verify the data types after standardizing date and time
df[['order_date', 'order_time']].dtypes

order_date    datetime64[ns]
order_time            object
dtype: object

In [7]:
# Sort orders chronologically by order date
df.sort_values(
    by='order_date',
    inplace=True,
    ascending=True
)

In [8]:
# Count unique values in each column and sort from least to most
df.nunique().sort_values()

return_status                    1
is_repeat_customer               2
customer_type                    3
review_sentiment                 3
gender                           3
customer_segment                 4
payment_status                   4
delivery_status                  4
shipping_method                  4
sales_channel                    4
order_status                     4
region                           5
currency                         7
customer_country                 7
payment_method                   7
return_reason                    8
marketing_channel               10
estimated_delivery_days         11
delivery_days                   15
campaign_name                   17
customer_order_count            19
warehouse                       19
quantity                        24
customer_review                 25
customer_rating                 35
customer_state                  38
coupon_code                     40
customer_age                    57
loyalty_points_redee

In [9]:
# Inspect unique values in categorical columns
categorical_columns = [
    'gender',
    'customer_segment',
    'customer_type',
    'region',
    'order_status',
    'sales_channel',
    'payment_method',
    'payment_status',
    'currency',
    'shipping_method',
    'warehouse',
    'delivery_status',
    'return_status',
    'return_reason',
    'review_sentiment',
    'marketing_channel',
    'is_repeat_customer'
]

for column in categorical_columns:
    print(f"\n{column}:")
    print(df[column].unique())


gender:
['Male' 'Female' 'Non-Binary']

customer_segment:
['Consumer' 'VIP' 'Business' 'Premium']

customer_type:
['Loyal' 'Returning' 'New']

region:
['North' 'South' 'Central' 'East' 'West']

order_status:
['Completed' 'Returned' 'Cancelled' 'Pending']

sales_channel:
['Website' 'Mobile App' 'Marketplace' 'Social Media']

payment_method:
['Credit Card' 'Debit Card' 'Digital Wallet' 'PayPal' 'Cash on Delivery'
 'Bank Transfer' 'Buy Now Pay Later']

payment_status:
['Paid' 'Refunded' 'Failed' 'Pending']

currency:
['AED' 'USD' 'GBP' 'CAD' 'AUD' 'INR' 'EUR']

shipping_method:
['Express' 'Economy' 'Same Day' 'Standard']

warehouse:
['WH-012' 'WH-016' 'WH-009' 'WH-007' 'WH-018' 'WH-015' 'WH-001' 'WH-002'
 'WH-008' 'WH-014' 'WH-004' 'WH-003' 'WH-006' 'WH-011' 'WH-017' 'WH-013'
 'WH-005' 'WH-019' 'WH-010']

delivery_status:
['On Time' 'Delayed' 'Cancelled' 'Early']

return_status:
[nan 'Returned']

return_reason:
[nan 'Late Delivery' 'Other' 'Product Not as Expected' 'Size Issue'
 'Damaged

In [10]:
# Summarize missing values by count and percentage
null_summary = pd.DataFrame({
    'Null Count': df.isnull().sum(),
    'Null Percentage': (df.isnull().sum() / len(df)) * 100
}).sort_values(by='Null Count', ascending=False)

null_summary

,Null Count,Null Percentage
return_reason,128654,93.149237
return_status,128654,93.149237
coupon_code,110502,80.006661
campaign_name,83233,60.263112
estimated_delivery_days,24557,17.779982
delivery_days,24557,17.779982
customer_rating,24557,17.779982
review_sentiment,24557,17.779982
customer_review,24557,17.779982
marketing_channel,0,0.000000


In [11]:
# Check order status for returned orders
df.loc[
    df['return_status'] == 'Returned',
    'order_status'
].value_counts(dropna=False)

order_status
Returned    9462
Name: count, dtype: int64

In [12]:
# Check order status for missing return statuses
df.loc[
    df['return_status'].isna(),
    'order_status'
].value_counts()

order_status
Completed    113559
Cancelled      8398
Pending        6697
Name: count, dtype: int64

In [13]:
# Check return reasons for orders with missing return status
df.loc[
    df['return_status'].isna(),
    'return_reason'
].value_counts(dropna=False)

return_reason
NaN    128654
Name: count, dtype: int64

In [14]:
# Check order status for orders with a return reason
df.loc[
    df['return_reason'].notna(),
    'order_status'
].value_counts(dropna=False)

order_status
Returned    9462
Name: count, dtype: int64

In [15]:
# Replace structural missing values with meaningful labels for non-returned orders
df['return_status'] = df['return_status'].fillna('Not Returned')
df['return_reason'] = df['return_reason'].fillna('No Return')

The missing values were identified as structural missingness because they represent orders with no return. Therefore, they were replaced with meaningful labels (Not Returned and No Return) to make the data easier to interpret during analysis.

In [16]:
# Confirm that no null values remain in the return fields
df[['return_status', 'return_reason']].isnull().sum()

return_status    0
return_reason    0
dtype: int64

In [17]:
# Compare discount amounts for orders with and without a coupon code
df.groupby(
    df['coupon_code'].isna()
)['discount_amount'].describe()

,count,mean,std,min,25%,50%,75%,max
coupon_code,,,,,,,,
False,27614.0,235.821261,374.906252,0.0,25.6700,103.275,278.8975,5046.32
True,110502.0,238.071515,379.741538,0.0,26.5325,103.285,276.8000,6759.02


In [18]:
# Check discount values for orders with missing coupon codes
df.loc[
    df['coupon_code'].isna(),
    'discount_amount'
].value_counts()

discount_amount
0.00       14236
10.89         14
1.19          14
4.80          13
1.99          13
           ...  
899.15         1
412.02         1
957.33         1
1275.08        1
1042.47        1
Name: count, Length: 51148, dtype: int64

In [19]:
# Check campaign names for orders with missing coupon codes
df.loc[
    df['coupon_code'].isna(),
    'campaign_name'
].value_counts(dropna=False)

campaign_name
NaN                   66747
Default_Campaign      18314
Google_Search_Q1       2226
Google_Shopping        2166
Google_Display         2166
FB_Retargeting         1828
Friend_Invite          1801
FB_Dynamic             1761
Referral_Program       1742
FB_Prospecting         1740
Insta_Influencer       1461
Insta_Reels            1429
Insta_Stories          1405
Abandoned_Cart         1220
Promo_Email            1177
Weekly_Newsletter      1160
Influencer_Program     1095
Affiliate_Partner      1064
Name: count, dtype: int64

In [20]:
# Check discount amounts for orders with a coupon code
df.loc[
    df['coupon_code'].notna(),
    ['coupon_code', 'discount_amount']
].groupby('coupon_code')['discount_amount'].agg(['count', 'mean', 'min', 'max']).sort_values('count', ascending=False)

,count,mean,min,max
coupon_code,,,,
SAVE39,731,248.723611,0.0,3448.46
SAVE49,727,229.139464,0.0,4463.91
SAVE40,719,250.095841,0.0,4748.59
SAVE12,716,238.572291,0.0,2836.67
SAVE33,715,241.835748,0.0,3771.54
SAVE26,712,239.620267,0.0,3057.22
SAVE25,711,255.223263,0.0,3187.41
SAVE15,709,243.847884,0.0,3227.35
SAVE36,709,247.265219,0.0,2887.83


In [21]:
# Compare whether discounts were applied to orders with missing coupon codes
df.groupby(
    df['coupon_code'].isna()
)['discount_amount'].agg(['count', 'sum', 'mean', 'min', 'max'])

,count,sum,mean,min,max
coupon_code,,,,,
False,27614,6511968.29,235.821261,0.0,5046.32
True,110502,26307378.53,238.071515,0.0,6759.02


Missing coupon_code values were retained as NaN because the available data does not provide sufficient evidence to determine the actual coupon code. Validation showed that orders with missing coupon codes can still receive substantial discounts and may be associated with marketing campaigns. Therefore, replacing the missing values with a label such as "No Coupon" could introduce inaccurate information.

In [22]:
# Check campaign names and missing values
df['campaign_name'].value_counts(dropna=False)

campaign_name
NaN                   83233
Default_Campaign      22989
Google_Search_Q1       2748
Google_Shopping        2746
Google_Display         2731
FB_Retargeting         2270
FB_Dynamic             2233
Friend_Invite          2232
Referral_Program       2198
FB_Prospecting         2161
Insta_Influencer       1822
Insta_Reels            1799
Insta_Stories          1773
Abandoned_Cart         1520
Promo_Email            1459
Weekly_Newsletter      1459
Influencer_Program     1404
Affiliate_Partner      1339
Name: count, dtype: int64

In [23]:
# Compare discounts across campaign names and missing values
df.groupby('campaign_name', dropna=False)['discount_amount'].agg(['count', 'mean', 'sum'])

,count,mean,sum
campaign_name,,,
Abandoned_Cart,1520,238.305178,362223.87
Affiliate_Partner,1339,248.893114,333267.88
Default_Campaign,22989,235.298641,5409280.45
FB_Dynamic,2233,249.385665,556878.19
FB_Prospecting,2161,232.444035,502311.56
FB_Retargeting,2270,232.495388,527764.53
Friend_Invite,2232,242.702110,541711.11
Google_Display,2731,235.778045,643909.84
Google_Search_Q1,2748,243.441656,668977.67


Missing campaign_name values were retained as NaN because the available data does not provide sufficient evidence that missing campaign names represent the absence of a campaign. Validation showed that records with missing campaign names still have substantial discount activity. Therefore, replacing missing values with "No Campaign" could introduce inaccurate information.

In [24]:
# Check the distribution and missing values of estimated delivery days
df['estimated_delivery_days'].value_counts(dropna=False).sort_index()

estimated_delivery_days
0.0      5794
1.0     15127
2.0      9380
3.0     19625
4.0     10247
5.0     14102
6.0     13871
7.0     13998
8.0      3855
9.0      3736
10.0     3824
NaN     24557
Name: count, dtype: int64

In [25]:
# Compare estimated delivery days with actual delivery days
df.groupby('estimated_delivery_days', dropna=False)['delivery_days'].agg(['count', 'mean', 'min', 'max'])

,count,mean,min,max
estimated_delivery_days,,,,
0.0,5794,0.367104,0.0,4.0
1.0,15127,1.290871,0.0,5.0
2.0,9380,2.244456,0.0,6.0
3.0,19625,3.238879,1.0,7.0
4.0,10247,4.260174,2.0,8.0
5.0,14102,5.246915,3.0,9.0
6.0,13871,6.249946,4.0,10.0
7.0,13998,7.252036,5.0,11.0
8.0,3855,8.243320,6.0,12.0


In [26]:
# Check delivery status for records with missing estimated delivery days
df.groupby('delivery_status', dropna=False)['estimated_delivery_days'].agg(['count', 'mean'])

,count,mean
delivery_status,,
Cancelled,0,NaN
Delayed,16886,4.305223
Early,9757,4.290766
On Time,86916,4.298553


In [27]:
# Check delivery status specifically for missing estimated delivery days
df[df['estimated_delivery_days'].isna()]['delivery_status'].value_counts(dropna=False)

delivery_status
Cancelled    24557
Name: count, dtype: int64

Missing estimated_delivery_days values were retained as NaN because all 24,557 missing values correspond to cancelled orders. Since cancelled orders did not proceed to delivery, an estimated delivery time is not applicable. Therefore, the missing values were not imputed.

In [28]:
# Check delivery days and missing values
df['delivery_days'].value_counts(dropna=False).sort_index()

delivery_days
0.0      6674
1.0     13001
2.0      9244
3.0     17221
4.0     10763
5.0     13988
6.0     13434
7.0     13216
8.0      5207
9.0      4717
10.0     4245
11.0      997
12.0      428
13.0      282
14.0      142
NaN     24557
Name: count, dtype: int64

In [29]:
# Check delivery status specifically for missing delivery days
df[df['delivery_days'].isna()]['delivery_status'].value_counts(dropna=False)

delivery_status
Cancelled    24557
Name: count, dtype: int64

Missing delivery_days values were retained as NaN because all 24,557 missing values correspond to cancelled orders. Since cancelled orders were not delivered, delivery duration is not applicable. Therefore, the missing values were not imputed.

In [30]:
# Check customer ratings and missing values
df['customer_rating'].value_counts(dropna=False).sort_index()

customer_rating
1.6        2
1.7        3
1.8        8
1.9       11
2.0       32
2.1       73
2.2      131
2.3      219
2.4      313
2.5      571
2.6      898
2.7     1288
2.8     1806
2.9     2527
3.0     3407
3.1     4476
3.2     5502
3.3     6618
3.4     7591
3.5     8755
3.6     9085
3.7     9362
3.8     9263
3.9     8744
4.0     7718
4.1     6628
4.2     5347
4.3     4177
4.4     3165
4.5     2197
4.6     1451
4.7      917
4.8      551
4.9      337
5.0      386
NaN    24557
Name: count, dtype: int64

In [31]:
# Check order status for records with missing customer ratings
df[df['customer_rating'].isna()]['delivery_status'].value_counts(dropna=False)

delivery_status
Cancelled    24557
Name: count, dtype: int64

Missing customer_rating values were retained as NaN because all 24,557 missing values correspond to cancelled orders. Since cancelled orders did not proceed to completion, a customer rating is not applicable. Therefore, the missing values were not imputed.

In [32]:
# Check review sentiment and missing values
df['review_sentiment'].value_counts(dropna=False)

review_sentiment
Positive    78083
Neutral     34684
NaN         24557
Negative      792
Name: count, dtype: int64

In [33]:
# Check delivery status for records with missing review sentiment
df[df['review_sentiment'].isna()]['delivery_status'].value_counts(dropna=False)

delivery_status
Cancelled    24557
Name: count, dtype: int64

Missing review_sentiment values were retained as NaN because all 24,557 missing values correspond to cancelled orders. Since cancelled orders did not generate customer reviews, sentiment is not applicable. Therefore, the missing values were not imputed.

In [34]:
# Check customer reviews and missing values
df['customer_review'].value_counts(dropna=False).head(10)

customer_review
NaN                                            24557
Good value for money.                          12237
Good product. Works as expected.               12152
Met my expectations. Would recommend.          12070
Satisfied with the purchase.                   11976
Product is okay, nothing special but works.    11954
Decent quality for the price.                  11855
Quality could be improved.                      6996
Product has some issues but it works.           6973
Mixed feelings about this purchase.             6929
Name: count, dtype: int64

In [35]:
# Check delivery status for records with missing customer reviews
df[df['customer_review'].isna()]['delivery_status'].value_counts(dropna=False)

delivery_status
Cancelled    24557
Name: count, dtype: int64

Missing customer_review values were retained as NaN because all 24,557 missing values correspond to cancelled orders. Since cancelled orders did not proceed to completion, no customer review was generated. Therefore, the missing values were not imputed.

In [36]:
# Deduplicate data and reset index
df = df.drop_duplicates().reset_index(drop=True)

In [37]:
# Export the cleaned dataset for analysis
df.to_csv('superstore_golden_dataset.csv', index=False)